app/src/main/java/com/example/tasky/MainActivity.kt

```diff
@@ -4,88 +4,112 @@ import android.os.Bundle
+import androidx.hilt.navigation.compose.hiltViewModel
+import androidx.navigation.NavHostController
+import androidx.navigation.compose.NavHost
+import androidx.navigation.compose.composable
+import androidx.navigation.compose.rememberNavController
 import com.example.tasky.authentication.presentation.login.LoginScreen
 import com.example.tasky.authentication.presentation.login.viewmodel.LoginEvent
 import com.example.tasky.authentication.presentation.login.viewmodel.LoginViewModel
+import com.example.tasky.authentication.presentation.signup.SignUpScreen
+import com.example.tasky.authentication.presentation.signup.viewmodel.SignUpEvent
+import com.example.tasky.authentication.presentation.signup.viewmodel.SignUpViewModel
 import com.example.tasky.ui.theme.TaskyTheme
 import dagger.hilt.android.AndroidEntryPoint
 
 @AndroidEntryPoint
 class MainActivity : ComponentActivity() {
 
-    private val loginViewModel by viewModels<LoginViewModel>()
     override fun onCreate(savedInstanceState: Bundle?) {
         super.onCreate(savedInstanceState)
         setContent {
             TaskyTheme {
-                Surface(
-                    modifier = Modifier.fillMaxSize(),
-                    color = MaterialTheme.colorScheme.background
-                ) {
-                    ObserveError()
-                    ObserveLogin()
-                    ObserveSignUp()
-                    LoginScreen(
-                        loginState = loginViewModel.state,
-                        onLoginEvent = loginViewModel::onEvent
-                    )
-                }
+                val navController = rememberNavController()
+                TaskyNavigation(navController)
             }
         }
     }
 
     @Composable
-    private fun ObserveError() {
-        LaunchedEffect(key1 = loginViewModel.state.errorMessage) {
-            if (loginViewModel.state.errorMessage != null) {
-                Toast.makeText(
-                    this@MainActivity,
-                    loginViewModel.state.errorMessage,
-                    Toast.LENGTH_LONG
-                ).show()
-                loginViewModel.onEvent(LoginEvent.ErrorShown)
+    private fun TaskyNavigation(navController: NavHostController) {
+        NavHost(navController = navController, startDestination = loginRoute) {
+            composable(loginRoute) {
+                val viewModel: LoginViewModel = hiltViewModel()
+                ObserveError(viewModel.state.errorMessage) {
+                    viewModel.onEvent(LoginEvent.ErrorShown)
+                }
+                ObserveBoolean(viewModel.state.onLoginSucceed) {
+                    viewModel.onEvent(LoginEvent.LoginNavigated)
+                    // TODO when home implemented
+                }
+                ObserveBoolean(viewModel.state.navigateToSignUp) {
+                    viewModel.onEvent(LoginEvent.SignUpNavigated)
+                    navController.navigate(signUpRoute)
+                }
```

Comment:
Can I handle the navigation in each screen also? Passing the navController as a parameter of the screen? Or can I retrieve the navController inside the screen?

Reply:
I wouldn´t send the navController to each screen, since you´ll give the screen too much permission to navigate to wherever they want, and can be messy

It´s better imo to pass a lambda

Ex: In the LoginScreen, send a Lambda for when the user clicks on "Dont have an account, create one", that redirects to the SignupScreen

```

@Composable
fun LoginScreen(state..event..., onSignup: () -> Unit)
```
and then wherever you instantiate the LoginScreen (TaskyNavigation) do:

```
...
LoginScreen(state...event.. onSignUp = {
navController.navigate("signup_screen")
}
)
```

Reply:
And in this case both screens have 2 navigation, should I create a lambda for each one? Or its better to create just 1 lambda with an event driven with sealed class logic?

Reply:
Both options are valid


In fact, I'll give you a third option
Add the event already on the LoginEvent, so you can use the same "onEvent" function



Tho, personally. I just use lambdas. I find them to be the easiest and fastest to do

Reply:
I'd always use callbacks/events to respond to user actions (such as button clicks). Passing down the nav controller can come with problems with the preview. 
I do something like this (simplified):
```
LoginScreen(
    ...,
    onEvent: (LoginEvent) -> Unit
) {
    Button(onClick = { onEvent(LoginEvent.Login) }) {
        Text("Login")
    }
}
```
and then in the composable block:
```
composable("login") {
    val viewModel = hiltViewModel<LoginViewModel>()
    LoginScreen(
        ...,
        onEvent = { event -> 
            when(event) {
                is LoginEvent.Login -> navController.navigate("agenda")
                else -> Unit
            }
            viewModel.onEvent(event)
        }
    )
}
```

Reply:
And if you need to do something, like navigate after a state change? ie Login succeed?

Reply:
```
composable("login") {
    val viewModel = hiltViewModel<LoginViewModel>()
    val state by viewModel.state.collectAsStateWithLifecycle()

    LaunchedEffect(state.isLoggedIn) {
        if(state.isLoggedIn) {
            navController.navigate("login") {
                popUpTo("login") {
                    inclusive = true
                }
            }
        }
    }

    LoginScreen(
        state = state,
        onEvent = { event -> 
            // ...
        }
    )
}
```